### PR TITLE
bulk-cdk: make exception classifiers recursive

### PR DESCRIPTION
**Pull request by postamar**
## What
I forgot to make the `ExceptionClassifier` implementations recursive, in terms of Exception causes. Without this, the `RegexExceptionClassifier` in particular is quite useless; because the interesting exceptions are usually wrapped inside other RuntimeExceptions and whatnot.

## How
Replicate the behaviour of the `ConnectorExceptionHandler` from the legacy CDK. 

## User Impact
None as of yet.

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/output/DefaultExceptionClassifier.kt [hunk 1]
```diff
@@ -22,24 +22,12 @@ class DefaultExceptionClassifier(
...
-        var connectorErrorException: ConnectorErrorException? = null
-        var unwound: Throwable? = e
-        while (unwound != null) {
-            if (unwound is ConnectorErrorException) {
-                connectorErrorException = unwound
-            }
-            unwound = unwound.cause
-        }
-        return connectorErrorException
-    }
```
Comment (postamar) [inline airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/output/DefaultExceptionClassifier.kt:44]: I generalized the unwinding logic, see below

File: airbyte-cdk/bulk/core/base/src/test/kotlin/io/airbyte/cdk/output/RegexExceptionClassifierTest.kt [hunk 2]
```diff
@@ -80,4 +80,16 @@ class RegexExceptionClassifierTest {
...
+    fun testRecursiveRuleOrdering() {
+        Assertions.assertEquals(
+            ConfigError("grouped: has foo\nhttps://www.youtube.com/watch?v=xvFZjo5PgG0"),
+            classifier.classify(RuntimeException("quux", RuntimeException("foobarbaz"))),
+        )
+        Assertions.assertEquals(
+            TransientError("barbaz"),
+            classifier.classify(RuntimeException("quux", RuntimeException("barbaz"))),
+        )
+    }
```
Comment (postamar) [inline airbyte-cdk/bulk/core/base/src/test/kotlin/io/airbyte/cdk/output/RegexExceptionClassifierTest.kt:94]: this is a regression test

File: airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/output/ExceptionClassifier.kt [hunk 3]
```diff
@@ -15,6 +15,16 @@ interface ExceptionClassifier : Ordered {
     val orderValue: Int
 
     override fun getOrder(): Int = orderValue
+
+    companion object {
+        fun unwind(e: Throwable, test: (Throwable) -> Boolean): Throwable? {
```
Comment (xiaohansong) [inline airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/output/ExceptionClassifier.kt:20]: optional - naming: test should be more like stopUnwind()

**Review by xiaohansong — APPROVED** [review]

File: airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/output/ExceptionClassifier.kt (hunk 3 above)
Comment (postamar) [reply airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/output/ExceptionClassifier.kt:20]: good idea!

Comment (postamar) [reply airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/output/ExceptionClassifier.kt]: done

**Pull request merged**